### ISSUE TITLE
Add commitlint for commit message linting

### ISSUE THREAD

**Issue by anikethsaha** [eslint/eslint#13124]
**The version of ESLint you are using.**
`7.0.0-alpha.3`

**The problem you want to solve.**
Easing the commit message formatting. Currently after committing and pushing it, In PR/push the commit messages gets checked and if there is an error it fails the CI.

**Your take on the correct solution to problem.**
Eslint does have githooks setup for linting the stage files.
We can use `commitlint` in order to lint the commit message in `commit-msg` git hook.
`commitlint` does provides rules configuration in order to write rules as per our needs and lint on that basis.

Packages needs to add (all devDep)

- `@commitlint/cli`
- `husky`
- `@commitlint/config-conventional` (optional if want to extend the base rules)`

**Rules which I think will satisfy eslint's format**

```json
  "commitlint": {
        "rules": {
            "header-max-length": [
                2,
                "always",
                72
            ],
            "scope-empty": [
                2,
                "always"
            ],
            "type-enum": [
                2,
                "always",
                [
                    "Fix",
                    "Update",
                    "New",
                    "Breaking",
                    "Docs",
                    "Build",
                    "Upgrade",
                    "Chore",
                    "Sponsors"
                ]
            ],
            "type-case": [
                2,
                "always",
                "start-case"
            ],
            "type-empty": [
                2,
                "never"
            ],
            "body-leading-blank": [
                1,
                "always"
            ]
        }
    },
```

**Screen Shot for wrong commit message**

![image](https://user-images.githubusercontent.com/26347874/77934626-f5726b00-72cd-11ea-9152-7e5bcf101fa7.png)

**Screenshot of correct commit message**

![image](https://user-images.githubusercontent.com/26347874/77934720-1dfa6500-72ce-11ea-94b2-2e02b562e660.png)

**Are you willing to submit a pull request to implement this change?**

Yes

**Comment by kaicataldo** [comment]
The commit message checker bot currently uses the 1st commit if there is a single commit and the PR title if there are multiple commits on the branch.

Is it possible to have it only check a specific commit on the branch (i.e. targetting the first commit that isn't in `master`)? Since we squash before merging, we currently have the ability to have separate commits in a PR, which can be very useful for reviewers.

The other thing that makes this tricky is if someone creates a branch off of another branch - how would `commitlint` know how which message to check? The first commit that doesn't match `master`'s history might be from another branch but is not part of the current branch's PR.

I would hate to force people to squash larger PRs into one commit to satisfy the linter (while making it more difficult to review). I suppose it could force each commit message to be more descriptive, though.

**Comment by anikethsaha** [comment]
I think it only lints the latest one of current branch.
I will have to confirm that though

**Comment by anikethsaha** [comment]
@kaicataldo I did some checks and it turns out 

> Is it possible to have it only check a specific commit on the branch (i.e. targetting the first commit that isn't in master)

committing only lint for the last commit on HEAD branch irrespective of the history. 

> The other thing that makes this tricky is if someone creates a branch off of another branch - how would `commitlint` know how which message to check? 


And it is triggered only during the commit-msg hook. And `commit-msg` triggers only after `pre-commit` so unless there are any staged changes, it won't trigger which is expected. 

> The first commit that doesn't match `master`'s history might be from another branch but is not part of the current branch's PR.

It doesn't look for the history, just the latest commit on HEAD

By latest commit, I meant the commit which is being processed and currently undergoing checks 

> I would hate to force people to squash larger PRs into one commit to satisfy the linter (while making it more difficult to review). I suppose it could force each commit message to be more descriptive, though.

I don't think it needs to be squashed for PR cause I think it generally not configured for CI. This is especially for the local machine
_And even if it's in CI (which is not needed as there is a bot for that), as long as the squash message covers the rules, it would be lint free_


Let me know if there is anything that I am missing. 👍

**Comment by kaicataldo** [comment]
In general, this seems like a good idea. One thing I’m concerned about is slowing down local development by forcing all commits in a PR to adhere to our commit message guidelines, which is extra overhead for the PR author. 

That being said, I think it would be great to have more helpful error messaging around commit message check failures, instead of failing the check without any insight as to why.

Another potential solution to this would be to [update our bot](https://github.com/eslint/eslint-github-bot/blob/master/src/plugins/commit-message/index.js) to display useful error messages to the user. I’m not sure what kind of messages can be shown in PR checks, but worst case scenario the bot could comment with a useful error message (and remove the comment when it passes).

**Comment by anikethsaha** [comment]
> One thing I’m concerned about is slowing down local development by forcing all commits in a PR to adhere to our commit message guidelines, which is extra overhead for the PR author.

we can use `--no-verify` and before merging if the last commit follows the format that's enough, it would clear the errors

> That being said, I think it would be great to have more helpful error messaging around commit message check failures, instead of failing the check without any insight as to why.


the one screenshot I shared seems to have a good error message. I don't know whether we can customize the message or not.


> Another potential solution to this would be to update our bot to display useful error messages to the user

This seems a good idea, but the bot runs in PRs or can say it runs on github (instead of local setup) and I was thinking of having commitlint in the local machine so that whenever we push wrong format commit, the bots fail the CI so we have to again amend and push. I wanted to eliminate this step by checking and linting the message before pushing.


Let me know if I am missing something

**Comment by nzakas** [comment]
Given that we check commit messages on pull requests, and we can always change the commit message when we merge a PR if it is incorrect, I'm not sure this is necessary. I'd rather people open a PR with an incorrect commit message that we fix manually upon merging instead of blocking people from opening a PR until they format the commit message correctly.

**Comment by anikethsaha** [comment]
>  I'd rather people open a PR with an incorrect commit message that we fix manually upon merging instead of blocking people from opening a PR until they format the commit message correctly.


This seems a good point. I do believe that commitlint's error messages are friendly to tackle the blocking thing and also they can use `--no-verify` as well.

**Comment by nzakas** [comment]
I’m not a fan of front-loading anything but the most important things into the precommit check. It’s already a bit slow and I’m not sure this is important enough to take another hit.

That said, I don’t feel strongly and am happy to go along with consensus if the team wants to move forward.
